examples/bioc/lib.rs

```diff
@@ -0,0 +1,157 @@
+    #[derive(scale::Encode, scale::Decode, PartialEq, Debug, Clone)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Status {
+        /// An auction has not started yet.
+        NotStarted,
+        /// We are in the starting period of the auction, collecting initial bids.
+        OpeningPeriod,
+        /// We are in the ending period of the auction, where we are taking snapshots of the winning
+        /// bids. Snapshots are taken currently on per-block basis, but this logic could be later evolve
+        /// to take snapshots of on arbitrary length (in blocks)
+        EndingPeriod(BlockNumber),
+        /// Candle was blown
+        Ended,
+        /// We have completed the bidding process and are waiting for the Random Function to return some acceptable
+        /// randomness to select the winner. The number represents how many blocks we have been waiting.
+        RfDelay(BlockNumber),
+    }
+
+    /// Struct for storing auction data.
+    #[derive(Debug, PartialEq, scale::Encode, scale::Decode, Clone)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct Auction {
+        /// Branded name of the auction event
+        name: String,
+        /// Some hash identifiyng the auction subject
+        subject: Hash,
+        /// Structure storing the bids being made
+        bids: Bids,
+        /// Auction terms encoded as:
+        /// [start_block, opening_period, closing_period]
+        terms: [BlockNumber; 3],
+        /// Auction status
+        status: Status,
+        /// Candle auction can have no winner.
+        /// If auction is finalized, that means that the winner is determined.
+        finalized: bool,
+    }
+
+    /// Way to fail a contract execution.
+    #[derive(scale::Encode, scale::Decode, Debug, PartialEq)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Failure {
+        Revert,
+        Panic,
+    }
+
+    impl Bioc {
+        #[ink(constructor)]
+        pub fn new() -> Self {
+            Self {}
+        }
+
+        /// Takes an auction data struct as input and returns it back.
+        #[ink(message)]
+        pub fn echo_auction(&mut self, auction: Auction) -> Auction {
+            auction
+        }
+
+        /// Fails contract execution in the required way.
+        #[ink(message)]
+        pub fn revert_or_trap(&mut self, fail: Option<Failure>) -> Result<(), Failure> {
+            match fail {
+                Some(Failure::Revert) => Err(Failure::Revert),
+                Some(Failure::Panic) => {
+                    panic!("Trapping on user demand!")
+                }
+                None => Ok(()),
+            }
+        }
+
+        /// Prints the specified string into node's debug log.
+        #[ink(message)]
+        pub fn debug_log(&mut self, _str: String) {
+            ink_env::debug_println!("debug_log: {}", _str);
+        }
```

Comment:
The message is used. Don't prefix it with `_`. The clippy error you get was because you used a keyword `str` as an identifier.

```suggestion
        pub fn debug_log(&mut self, message: String) {
            ink_env::debug_println!("debug_log: {}", message);
        }
```